reliability/single-fit/okhttp/src/test/kotlin/com/datadog/android/okhttp/RumContextPropagationTest.kt

```diff
@@ -0,0 +1,278 @@
+import com.datadog.android.core.sampling.DeterministicSampler.Companion.SAMPLER_HASHER
+import com.datadog.android.core.stub.StubSDKCore
+import com.datadog.android.okhttp.tests.elmyr.OkHttpConfigurator
+import com.datadog.android.okhttp.trace.TracingInterceptor
+import com.datadog.android.trace.DatadogTracing
+import com.datadog.android.trace.GlobalDatadogTracer
+import com.datadog.android.trace.Trace
+import com.datadog.android.trace.TraceConfiguration
+import com.datadog.android.trace.TracingHeaderType
+import com.datadog.android.trace.api.TestIdGenerationStrategy
+import com.datadog.android.trace.api.setTestIdGenerationStrategy
+import com.datadog.android.trace.api.tracer.DatadogTracerBuilder
+import com.datadog.tools.unit.extensions.TestConfigurationExtension
+import com.datadog.tools.unit.getFieldValue
+import com.datadog.tools.unit.getStaticValue
+import fr.xgouchet.elmyr.Forge
+import fr.xgouchet.elmyr.junit5.ForgeConfiguration
+import fr.xgouchet.elmyr.junit5.ForgeExtension
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.mockwebserver.MockResponse
+import okhttp3.mockwebserver.MockWebServer
+import okhttp3.mockwebserver.RecordedRequest
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.AfterEach
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.extension.ExtendWith
+import org.junit.jupiter.api.extension.Extensions
+import org.mockito.junit.jupiter.MockitoExtension
+import org.mockito.junit.jupiter.MockitoSettings
+import org.mockito.quality.Strictness
+
+@Extensions(
+    ExtendWith(MockitoExtension::class),
+    ExtendWith(ForgeExtension::class),
+    ExtendWith(TestConfigurationExtension::class)
+)
+@ForgeConfiguration(OkHttpConfigurator::class)
+@MockitoSettings(strictness = Strictness.LENIENT)
+class RumContextPropagationTest {
+    private lateinit var mockServer: MockWebServer
+    private lateinit var testedClient: OkHttpClient
+    private lateinit var stubSdkCore: SdkCore
+
+    @BeforeEach
+    fun `set up`() {
+        mockServer = MockWebServer().apply {
+            enqueue(MockResponse())
+            start()
+        }
+    }
+
+    @AfterEach
+    fun `tear down`() {
+        Datadog.stopInstance(stubSdkCore.name)
+        mockServer.shutdown()
+    }
+
+    @Test
+    fun `M send rum sessionId in baggage header W call is made`(forge: Forge) {
+        // Given
+        val fakeRumContext = forge.createRumContext(SAMPLED_IDS.random())
+        stubSdkCore = forge.prepareSubSdkCore(fakeRumContext)
+        testedClient = prepareClient(stubSdkCore)
+        registerTracer(createTracer(stubSdkCore))
+
+        // When
+        testedClient.makeNetworkCall()
+
+        // Then
+        assertSentRequest {
+            assertThat(getHeader(HEADER_BAGGAGE)).isEqualTo(
+                "session.id=${fakeRumContext[RUM_CONTEXT_SESSION_ID]}"
+            )
+        }
```

Comment:
Most probably all these tests will fail if you try to run them with a real SDK core which have a dedicated thread for context processing, because of the concurrency issue described above.

Reply:
I'll keep this tests as it at least test the case when rum context is resolved. 

The rest of the cases is covered in a different places (I reverted the lazy context resulution logic back with slight refactor).
 
Note that I didn't add any synchronization for the RUM context future before data is being sent by span event or http header, but added MAINTEINER logs level to see if we have such issue.